### ISSUE TITLE
Cpptools return wrong locations when using command "vscode.executeReferenceProvider"

### ISSUE THREAD

**Issue by Lennon925** [microsoft/vscode-cpptools#4147]
Cpptools Version: 0.25

Currently cpptools return wrong locations when using command "vscode.executeReferenceProvider" and passing uri and position to this command. It is using current selected word position of current editor instead of uri and  position I passed to. You can repro this like this:

let uri = xxxxx;
let position = xxx;
vscode.commands.executeCommand<vscode.Location[]>('vscode.executeReferenceProvider', uri, position).then(locations => {
//do something.
}

Select different word of current editor and run codes above and you will get differents locations results.

**Comment by sean-mcmanus** [comment]
I tried the repro, but I'm not seeing a bug. Can you provide more repro info? I see the code using the initial URI and position, but if a different query is made while the previous one is still processing it will trigger a preview of the initial request (ignoring the URL/position of the new cursor position), which is currently "by design". The code I used to test this is
```
    private prevUri: vscode.Uri;
    private prevPosition: vscode.Position;
    public selectionChanged(selection: Range): void {
        this.notifyWhenReady(() => {
            let uri: vscode.Uri = this.activeDocument.uri;
            let position: vscode.Position = new vscode.Position(selection.start.line, selection.start.character);
            if (!this.prevUri) {
                this.prevUri = uri;
                this.prevPosition = position;
            }
            vscode.commands.executeCommand<vscode.Location[]>('vscode.executeReferenceProvider', this.prevUri, this.prevPosition).then(locations => {
                if (locations.length > 0) {
                    console.log("line: " + locations[0].range.start.line);
                }
            });
            this.prevUri = uri;
            this.prevPosition = position;
            this.languageClient.sendNotification(TextEditorSelectionChangeNotification, selection);
        });
    }
```
Which correctly causes the previous cursor position to be used for references and not the current position.

**Comment by Lennon925** [comment]
> I tried the repro, but I'm not seeing a bug. Can you provide more repro info? I see the code using the initial URI and position, but if a different query is made while the previous one is still processing it will trigger a preview of the initial request (ignoring the URL/position of the new cursor position), which is currently "by design". The code I used to test this is
> 
> ```
>     private prevUri: vscode.Uri;
>     private prevPosition: vscode.Position;
>     public selectionChanged(selection: Range): void {
>         this.notifyWhenReady(() => {
>             let uri: vscode.Uri = this.activeDocument.uri;
>             let position: vscode.Position = new vscode.Position(selection.start.line, selection.start.character);
>             if (!this.prevUri) {
>                 this.prevUri = uri;
>                 this.prevPosition = position;
>             }
>             vscode.commands.executeCommand<vscode.Location[]>('vscode.executeReferenceProvider', this.prevUri, this.prevPosition).then(locations => {
>                 if (locations.length > 0) {
>                     console.log("line: " + locations[0].range.start.line);
>                 }
>             });
>             this.prevUri = uri;
>             this.prevPosition = position;
>             this.languageClient.sendNotification(TextEditorSelectionChangeNotification, selection);
>         });
>     }
> ```
> 
> Which correctly causes the previous cursor position to be used for references and not the current position.

I mean that it will always use activeDocument.uri and selection position to call command "vscode.executeReferenceProvider" instead of params I pass to it.  The uri and position I passed maybe is not equal the activeDocument uri and selection position.

let uri = xxxxx;  //(test.cpp)
let position = xxx;  //(line 5,column 6)
vscode.commands.executeCommand<vscode.Location[]>('vscode.executeReferenceProvider', uri, position).then(locations => {
//do something.
}
if current selection is line 5 column 6 and this command will return correct results. But after changing the selection to line 100, column 20, run this command and pass position line 5, column 6, this command will return wrong results. Because it use position line 100 and column 20.

**Comment by sean-mcmanus** [comment]
In my repro code example, it is using the args passed to it and not the selection position.

I changed my repro code to use a hardcoded URI and position and it uses that value and not the active doc or selection.

Our language server should just be receiving the params that are passed -- we have absolutely no code that checks the active document or selection. My guess is that your code is somehow incorrect?

**Comment by Lennon925** [comment]
> In my repro code example, it is using the args passed to it and not the selection position.
> 
> I changed my repro code to use a hardcoded URI and position and it uses that value and not the active doc or selection.
> 
> Our language server should just be receiving the params that are passed -- we have absolutely no code that checks the active document or selection. My guess is that your code is somehow incorrect?

I think my code is right and every time I can repro this bug. I selected the first "Age" marked in picture and pass the first "Age" position to command and run , result is as the first log.
I selected the second "Age" and pass the first "Age" position to command and run, result is as the second log.

![捕获](https://user-images.githubusercontent.com/39362073/63829806-16bc5300-c99d-11e9-87bd-e1ccc2a11462.PNG)
![捕获](https://user-images.githubusercontent.com/39362073/63829759-f391a380-c99c-11e9-9bb6-9caa475c4ba1.PNG)

**Comment by Lennon925** [comment]
`void Test8()
{
    Student *student = new Student();
    student->Gender = GENDER::FEMALE;
}`

**Comment by Lennon925** [comment]
In output window，c/c++ references output can also support this. It lists the references of the word you selected currently instead of the word you passed to command "vscode.executeReferenceProvider".

my code is as below and If you selected different words, the length of locations are different.

let uri: vscode.Uri = vscode.Uri.file('d:\\Projects\\demo\\Case\\test.cpp');
let pos: vscode.Position = new vscode.Position(40, 14);
return vscode.commands.executeCommand<vscode.Location[]>('vscode.executeReferenceProvider', uri, pos).then(locations => {
			console.log('**************** ' + locations.length + '\n');
}

**Comment by Lennon925** [comment]
I don't know if it's related with this error: every time cpptools launch, "C/C++: Install vsix -- Failed to download rate limit JSON " occurs.

**Comment by sean-mcmanus** [comment]
The "download rate limit JSON" error can be ignored -- it's just from code that checks if there's a new  Insiders version available.

I tried your repro using the same code in the screenshot and I'm still getting results that are the same no matter what the cursor position is.

Are you sure you're using the latest version of our extension code from 0.25.0 and there are no other reference providers?

**Comment by sean-mcmanus** [comment]
In your repro, are you selecting code or lines of code when you invoke the reference provider? I'm seeing a bug when it's invoked with certain code selected, but not when the cursor is moved without any selection.

**Comment by sean-mcmanus** [comment]
Yeah, the bug was with selection -- we'll see if we can get this into 0.25.1.

**Comment by Lennon925** [comment]
> Yeah, the bug was with selection -- we'll see if we can get this into 0.25.1.

Thanks

**Comment by sean-mcmanus** [comment]
This should be fixed with 0.25.1.
